### PR TITLE
Cpt 835 terraform cloud

### PR DESCRIPTION
**Pull request by EmmanuelOgiji**
- Add helm chart support for terraform cloud workspace
- Add toggle to standard-application-chart helm chart to allow gradual migration of external secrets

### CODE REVIEW

File: charts/terraform-cloud/templates/_helpers.tpl [hunk 1]
```diff
@@ -0,0 +1,98 @@
+{{- define "mintel_common.workspace" }}
```
Comment (nabadger) [inline charts/terraform-cloud/templates/_helpers.tpl:1]: The new style of doing helpers is to not bundle them into 1 huge file, and instead create more specific files under `./helpers` dir

File: charts/terraform-cloud/templates/output-secret.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{{- if .Values.global.terraform.externalSecrets }}
```
Comment (nabadger) [inline charts/terraform-cloud/templates/output-secret.yaml:1]: name of this file is too generic I think (it's tf-workspace specific)

File: charts/terraform-cloud/templates/_helpers.tpl [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+kind: Workspace
+metadata:
+  name: "{{ $global.clusterEnv }}-{{ $global.clusterRegion }}-{{ $global.clusterName }}-{{ $global.namespace }}-{{ .name }}-{{ $resourceType | kebabcase }}"
+  namespace: {{ $.Release.Namespace | quote }}
+  labels: {{ include "mintel_common.labels" $ | nindent 4 }}
+  annotations:
+    {{ include "mintel_common.commonAnnotations" $ | nindent 4 }}
+    argocd.argoproj.io/sync-options: SkipDryRunOnMissingResource=true
+spec:
+  agentPoolID: {{ has $global.clusterEnv (list "prod" "logs") | ternary "apool-RhENdyZD1fV8Kdde" "apool-ARFKgcQQcY3T91bk" | quote }}
```
Comment (barrymars) [inline charts/terraform-cloud/templates/_helpers.tpl:19]: would be good to have these real values pulled out of the chart....maybe populated by the jsonnet wrapper, or ArgoCD

Comment (EmmanuelOgiji) [reply charts/terraform-cloud/templates/_helpers.tpl:19]: That makes sense but it's also a weird one where it's used in just one place so is that worth the potential hassle maybe

**Comment by nabadger** [comment]
Is there a benefit to naming the input vars `defaultVars` vs just `vars` - i.e. are there non-default vars? :) 

Picky, but it's just more to type :)

File: charts/terraform-cloud/values.yaml [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+  # -- Kubernetes cluster region
+  clusterRegion: ""
+  # -- Global variables relating to cloud provider
+  cloudProvider:
+    # -- AWS account ID
+    accountId: ""
+  # -- Global variable definint RUNTIME_ENVIRONMENT
+  runtimeEnvironment: kubernetes
+  # -- Global dictionary of TLS secrets
+  ingressTLSSecrets: {}
```
Comment (nabadger) [inline charts/terraform-cloud/values.yaml:29]: Are these used?

**Comment by EmmanuelOgiji** [comment]
> 

@nabadger We built in the ability to render say multiple instances of resources. So defaultVars will be what you share across all the instances for example

**Comment by bcbrockway** [comment]
> @nabadger We built in the ability to render say multiple instances of resources. So defaultVars will be what you share across all the instances for example

@EmmanuelOgiji @nabadger I'm adding more stuff to the values file / docs to make this clearer.

File: charts/terraform-cloud/templates/helpers/_workspace.yaml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+kind: Workspace
+metadata:
+  name: "{{ $global.clusterEnv }}-{{ $global.clusterRegion }}-{{ $global.clusterName }}-{{ $.Release.Namespace }}-{{ .name }}-{{ $resourceType | kebabcase }}"
+  namespace: {{ $.Release.Namespace | quote }}
+  labels: {{ include "mintel_common.labels" $ | nindent 4 }}
+  annotations:
+    {{ include "mintel_common.commonAnnotations" $ | nindent 4 }}
+    argocd.argoproj.io/sync-options: SkipDryRunOnMissingResource=true
+spec:
+  agentPoolID: {{ has $global.clusterEnv (list "prod" "logs") | ternary "apool-RhENdyZD1fV8Kdde" "apool-ARFKgcQQcY3T91bk" | quote }}
```
[the remark this answers is not in the document]
Comment (EmmanuelOgiji) [reply charts/terraform-cloud/templates/helpers/_workspace.yaml:19]: The thing is the pools should almost never change unless we change the architecture. The other options like Argo would mean that we still have to hard code in a more complex way. Then we can say read it from tf cloud or say secrets manager but then that means people have to set up credentials before rendering manifests. I'm not sure we want that

**Review by bcbrockway — APPROVED** [review]

**Pull request merged**